Changes in module 'nf-core/bakta/bakta'
--- /dev/null
+++ modules/nf-core/bakta/bakta/modules.json
@@ -0,0 +1,5 @@
+{
+  "name": "",
+  "homePage": "",
+  "repos": {}
+}

--- modules/nf-core/bakta/bakta/main.nf
+++ modules/nf-core/bakta/bakta/main.nf
@@ -10,8 +10,6 @@
     input:
     tuple val(meta), path(fasta)
     path db
-    path proteins
-    path prodigal_tf
 
     output:
     tuple val(meta), path("${prefix}.embl")             , emit: embl
@@ -24,6 +22,8 @@
     tuple val(meta), path("${prefix}.hypotheticals.faa"), emit: hypotheticals_faa
     tuple val(meta), path("${prefix}.tsv")              , emit: tsv
     tuple val(meta), path("${prefix}.txt")              , emit: txt
+    tuple val(meta), path("${prefix}.svg")              , emit: svg
+    tuple val(meta), path("${prefix}.png")              , emit: png
     path "versions.yml"                                 , emit: versions
 
     when:
@@ -32,17 +32,21 @@
     script:
     def args = task.ext.args   ?: ''
     prefix   = task.ext.prefix ?: "${meta.id}"
-    def proteins_opt = proteins ? "--proteins ${proteins[0]}" : ""
-    def prodigal_tf = prodigal_tf ? "--prodigal-tf ${prodigal_tf[0]}" : ""
     """
     bakta \\
         $fasta \\
         $args \\
         --threads $task.cpus \\
         --prefix $prefix \\
-        $proteins_opt \\
-        $prodigal_tf \\
-        --db $db
+        --locus-tag $prefix \\
+        --db $db \\
+        --keep-contig-headers \\
+        --skip-trna \\
+        --skip-tmrna \\
+        --skip-rrna \\
+        --skip-ncrna \\
+        --skip-ncrna-region \\
+        --skip-crispr
 
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":

************************************************************
