Changes in module 'nf-core/gecco/run'
--- modules/nf-core/gecco/run/main.nf
+++ modules/nf-core/gecco/run/main.nf
@@ -13,12 +13,12 @@


     output:
-    tuple val(meta), path("*.genes.tsv")    , optional: true, emit: genes
-    tuple val(meta), path("*.features.tsv")                 , emit: features
-    tuple val(meta), path("*.clusters.tsv") , optional: true, emit: clusters
+    tuple val(meta), path("*.genes.tsv"),     optional: true, emit: genes
+    tuple val(meta), path("*.features.tsv"),                  emit: features
+    tuple val(meta), path("*.clusters.tsv"),  optional: true, emit: clusters
     tuple val(meta), path("*_cluster_*.gbk"), optional: true, emit: gbk
-    tuple val(meta), path("*.json")         , optional: true, emit: json
-
+    tuple val(meta), path("*clusters*.gff"), optional: true, emit: gff
+    tuple val(meta), path("*.json"),          optional: true, emit: json
     path "versions.yml"                     , emit: versions

     when:
@@ -39,6 +39,8 @@
         $custom_model \\
         $custom_hmm

+    gecco convert clusters -i ./ --format gff
+
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":
         gecco: \$(echo \$(gecco --version) | cut -f 2 -d ' ' )

************************************************************
